### PR TITLE
Added docblock in rclcpp

### PR DESCRIPTION
**Pull request by ahcorde**
Added docblock to some undocumented classes and methods in rclcpp

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: rclcpp/include/rclcpp/client.hpp [hunk 1]
```diff
@@ -84,22 +84,42 @@ class ClientBase
   bool
   take_type_erased_response(void * response_out, rmw_request_id_t & request_header_out);
 
+  /// Return the name of the node.
```
Comment (brawner) [inline rclcpp/include/rclcpp/client.hpp:87]: name of the service

File: rclcpp/include/rclcpp/duration.hpp [hunk 2]
```diff
@@ -26,11 +26,22 @@ namespace rclcpp
 class RCLCPP_PUBLIC Duration
 {
 public:
+  /// Duration constructor.
+  /**
+   * Initializes the time values for seconds and nanoseconds individually.
+   * Large values for nsecs are wrapped automatically with the remainder added to secs.
+   * Both inputs must be integers.
+   * Durations can be negative.
```
Comment (brawner) [inline rclcpp/include/rclcpp/duration.hpp:34]: Really, just the seconds component can be negative, nanoseconds is an unsigned int

File: rclcpp/include/rclcpp/duration.hpp [hunk 3]
```diff
@@ -26,11 +26,22 @@ namespace rclcpp
...
+   *
+   * \param seconds time in seconds
```
Comment (brawner) [inline rclcpp/include/rclcpp/duration.hpp:36]: Fix these two parameter descriptions. It's a duration, and the nanoseconds component says seconds.

File: rclcpp/include/rclcpp/duration.hpp [hunk 4]
```diff
@@ -44,69 +55,98 @@ class RCLCPP_PUBLIC Duration
...
+   */
   explicit Duration(const rcl_duration_t & duration);
 
+  /// Copy constructor.
   Duration(const Duration & rhs);
 
+  /// Duration destructor.
   virtual ~Duration() = default;
 
+  /// Return a builtin_interfaces::msg::Duration object based
```
Comment (brawner) [inline rclcpp/include/rclcpp/duration.hpp:70]: Incomplete sentence

File: rclcpp/include/rclcpp/duration.hpp [hunk 5]
```diff
@@ -44,69 +55,98 @@ class RCLCPP_PUBLIC Duration
...
   operator builtin_interfaces::msg::Duration() const;
 
   // cppcheck-suppress operatorEq // this is a false positive from cppcheck
   Duration &
   operator=(const Duration & rhs);
 
+  /// Copy assign operator from a builtin_interfaces::msg::Duration.
```
Comment (brawner) [inline rclcpp/include/rclcpp/duration.hpp:77]: I don't think the operator comments below necessary? You're not really saying anything informative

File: rclcpp/include/rclcpp/duration.hpp [hunk 6]
```diff
@@ -44,69 +55,98 @@ class RCLCPP_PUBLIC Duration
...
 
+  /// Addition operator.
   Duration
   operator+(const rclcpp::Duration & rhs) const;
 
+  /// Subtraction operator.
   Duration
   operator-(const rclcpp::Duration & rhs) const;
 
+  /// Get the maximum duration value.
```
Comment (brawner) [inline rclcpp/include/rclcpp/duration.hpp:109]: Is this the maximum representable value?

File: rclcpp/include/rclcpp/duration.hpp [hunk 7]
```diff
@@ -44,69 +55,98 @@ class RCLCPP_PUBLIC Duration
...
+  /**
+   * \return the maximum duration value
+   */
   static
   Duration
   max();
 
+  /// Multiplication operator.
```
Comment (brawner) [inline rclcpp/include/rclcpp/duration.hpp:117]: Unnecessary?

File: rclcpp/include/rclcpp/node.hpp [hunk 8]
```diff
@@ -229,15 +229,28 @@ class Node : public std::enable_shared_from_this<Node>
...
+   */
   template<typename ServiceT>
   typename rclcpp::Client<ServiceT>::SharedPtr
   create_client(
     const std::string & service_name,
     const rmw_qos_profile_t & qos_profile = rmw_qos_profile_services_default,
     rclcpp::CallbackGroup::SharedPtr group = nullptr);
 
-  /* Create and return a Service. */
+  /* Create and return a Service.
```
Comment (brawner) [inline rclcpp/include/rclcpp/node.hpp:246]: Use /// for the brief.

File: rclcpp/include/rclcpp/parameter.hpp [hunk 9]
```diff
@@ -117,50 +121,62 @@ class Parameter
   decltype(auto)
   get_value() const;
 
+  /// Get value of parameter as boolean.
```
Comment (brawner) [inline rclcpp/include/rclcpp/parameter.hpp:124]: Do these throw if they are the wrong type? I would add that in here. Same for all the as_* methods below.

File: rclcpp/include/rclcpp/service.hpp [hunk 10]
```diff
@@ -50,14 +50,26 @@ class ServiceBase
   RCLCPP_PUBLIC
   virtual ~ServiceBase();
 
+  /// Return the name of the node.
```
Comment (brawner) [inline rclcpp/include/rclcpp/service.hpp:53]: Fix these two lines.

File: rclcpp/include/rclcpp/service.hpp [hunk 11]
```diff
@@ -50,14 +50,26 @@ class ServiceBase
...
+  /** \return The name of the node. */
   RCLCPP_PUBLIC
   const char *
   get_service_name();
 
+  /// Return the rcl_client_t client handle in a std::shared_ptr.
```
Comment (brawner) [inline rclcpp/include/rclcpp/service.hpp:59]: Fix the remaining comments, they all refer to clients.

File: rclcpp/include/rclcpp/subscription_options.hpp [hunk 12]
```diff
@@ -89,6 +89,10 @@ struct SubscriptionOptionsWithAllocator : public SubscriptionOptionsBase
   {}
 
   /// Convert this class, with a rclcpp::QoS, into an rcl_subscription_options_t.
+  /**
+   * \param qos QoS profile.
```
Comment (brawner) [inline rclcpp/include/rclcpp/subscription_options.hpp:93]: Is there anything more you can say?

File: rclcpp/include/rclcpp/time.hpp [hunk 13]
```diff
@@ -31,96 +31,149 @@ class Clock;
 class Time
 {
 public:
+  /// Time constructor
+  /**
+   * \param seconds since time epoch
+   * \param nanoseconds since time epoch
```
Comment (brawner) [inline rclcpp/include/rclcpp/time.hpp:37]: Fix

File: rclcpp/include/rclcpp/time.hpp [hunk 14]
```diff
@@ -31,96 +31,149 @@ class Clock;
...
+   * \param clock_type clock type
+   */
   RCLCPP_PUBLIC
   explicit Time(const rcl_time_point_t & time_point);
 
+  /// Time destructor
   RCLCPP_PUBLIC
   virtual ~Time();
 
+  /// Return a builtin_interfaces::msg::Time object based
```
Comment (brawner) [inline rclcpp/include/rclcpp/time.hpp:77]: incomplete sentence

File: rclcpp/include/rclcpp/time.hpp [hunk 15]
```diff
@@ -31,96 +31,149 @@ class Clock;
...
   RCLCPP_PUBLIC
   operator builtin_interfaces::msg::Time() const;
 
+  /// Copy assign operator
```
Comment (brawner) [inline rclcpp/include/rclcpp/time.hpp:81]: These operator comments aren't very descriptive. Either add more information, or just leave them out.

File: rclcpp/include/rclcpp/time.hpp [hunk 16]
```diff
@@ -31,96 +31,149 @@ class Clock;
...
   operator+=(const rclcpp::Duration & rhs);
 
+  /// Subtraction assigment operator with a rclcpp::Duration.
   RCLCPP_PUBLIC
   Time &
   operator-=(const rclcpp::Duration & rhs);
 
+  /// Get the nanoseconds since epoch
+  /**
+   * \return the nanoseconds since epoch as a floating point number.
```
Comment (brawner) [inline rclcpp/include/rclcpp/time.hpp:148]: It is not a floating point number.

File: rclcpp/include/rclcpp/time.hpp [hunk 17]
```diff
@@ -31,96 +31,149 @@ class Clock;
...
+   */
   RCLCPP_PUBLIC
   rcl_time_point_value_t
   nanoseconds() const;
 
+  /// Get the maximum time value
```
Comment (brawner) [inline rclcpp/include/rclcpp/time.hpp:154]: Maximum possible representable time value?

File: rclcpp/include/rclcpp/time_source.hpp [hunk 18]
```diff
@@ -54,19 +81,24 @@ class TimeSource
     rclcpp::node_interfaces::NodeClockInterface::SharedPtr node_clock_interface,
     rclcpp::node_interfaces::NodeParametersInterface::SharedPtr node_parameters_interface);
 
+  /// Detach the node from the tiem source
```
Comment (brawner) [inline rclcpp/include/rclcpp/time_source.hpp:84]:
```suggestion
  /// Detach the node from the time source
```

File: rclcpp/include/rclcpp/time_source.hpp [hunk 19]
```diff
@@ -54,19 +81,24 @@ class TimeSource
...
   RCLCPP_PUBLIC
   void detachNode();
 
   /// Attach a clock to the time source to be updated
   /**
+   * The time source must be a RCL_ROS_TIME otherwise throws an exception
```
Comment (brawner) [inline rclcpp/include/rclcpp/time_source.hpp:90]: Instead, add another /throws line.

File: rclcpp/include/rclcpp/time_source.hpp [hunk 20]
```diff
@@ -54,19 +81,24 @@ class TimeSource
...
    * \throws std::invalid_argument if node is nullptr
    */
   RCLCPP_PUBLIC
   void attachClock(rclcpp::Clock::SharedPtr clock);
 
+  /// Detach a clock to the time source
   RCLCPP_PUBLIC
   void detachClock(rclcpp::Clock::SharedPtr clock);
 
+  /// TimeSource Destuctor
```
Comment (brawner) [inline rclcpp/include/rclcpp/time_source.hpp:101]:
```suggestion
  /// TimeSource Destructor
```

File: rclcpp/include/rclcpp/timer.hpp [hunk 21]
```diff
@@ -71,10 +82,15 @@ class TimerBase
...
 
+  /// Reset the timer.
+  /**
+   * \throws std::runtime_error if the rcl_timer_reset returns a failure
+   */
   RCLCPP_PUBLIC
   void
   reset();
 
+  /// Handles the function to execute when the timer is emit a trigger.
```
Comment (brawner) [inline rclcpp/include/rclcpp/timer.hpp:93]: This calls the callback function when the timer signal is emitted right? It doesn't handle it

File: rclcpp/include/rclcpp/timer.hpp [hunk 22]
```diff
@@ -169,6 +185,10 @@ class GenericTimer : public TimerBase
     cancel();
   }
 
+  /// Handles the function to execute when the timer is emit a trigger.
```
Comment (brawner) [inline rclcpp/include/rclcpp/timer.hpp:188]: These are derived functions. You don't need to re-document them. I'm pretty sure, doxygen will carry forward the inherited documentation.

File: rclcpp/include/rclcpp/duration.hpp [hunk 23]
```diff
@@ -80,33 +95,46 @@ class RCLCPP_PUBLIC Duration
...
+   */
   double
   seconds() const;
 
-  // Create a duration object from a floating point number representing seconds
+  /// Create a duration object from a floating point number representing seconds
   static Duration
   from_seconds(double seconds);
 
+  /// Convert Duration in to std::chrono.
```
Comment (brawner) [inline rclcpp/include/rclcpp/duration.hpp:129]:
```suggestion
  /// Convert Duration into a std::chrono::Duration.
```

**Comment by ahcorde** [comment]
Running CI up-to rclcpp to check linters

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10628/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6062/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8655/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10534/)

**Comment by ahcorde** [comment]
The CI warnings on mac are not related to linters which is the only thing that this PR may affect because it just documentation. Merging it.

**Pull request merged**